util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+        HANDLE filemap = CreateFileMapping(handle, NULL, PAGE_READONLY, high, low, NULL);
+        void* base = NULL;
+        if (filemap == NULL) {
+          s = WindowsError(fname, GetLastError());
+        } else {
+          base = MapViewOfFile(
+            filemap,
+            FILE_MAP_READ,
+            0,
+            0,
+            size
+          );
+        }
+
+        if (base != NULL) {
+          *result = new WindowsMmapReadableFile(fname, base, size, filemap, &mmap_limit_);
+        } else {
+          CloseHandle(filemap);
+          s = WindowsError(fname, GetLastError());
+        }
+      }
+      CloseHandle(handle);
+      if (!s.ok()) {
+        mmap_limit_.Release();
+      }
+    } else {
+      *result = new WindowsRandomAccessFile(fname, handle, &fd_limit_);
+    }
+    return s;
+  }
+
+  virtual Status NewWritableFile(const std::string& fname,
+                                 WritableFile** result) {
+    HANDLE h = CreateFileW(
+      Utf8ToUtf16(fname).data(),
+      GENERIC_WRITE,
+      // Use shared read/write since corruption_test requires
+      // access to a concurrently open logger, opened with shared
+      // "w+" access.
+      FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+      NULL,
+      CREATE_ALWAYS,
+      FILE_ATTRIBUTE_NORMAL,
+      NULL
+    );
+
+    if (h == INVALID_HANDLE_VALUE) {
+      *result = NULL;
+      return WindowsError(fname, GetLastError());
+    } else {
+      *result = new WindowsWritableFile(fname, h);
+      return Status::OK();
+    }
+  }
+
+  virtual Status NewAppendableFile(const std::string& fname,
+                                   WritableFile** result) {
+    HANDLE h = CreateFileW(
+      Utf8ToUtf16(fname).data(),
+      FILE_APPEND_DATA,
+      // Use shared read/write since corruption_test requires
+      // access to a concurrently open logger, opened with shared
+      // "w+" access.
+      FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+      NULL,
+      OPEN_ALWAYS,
+      FILE_ATTRIBUTE_NORMAL,
+      NULL
+    );
+    if (h == INVALID_HANDLE_VALUE) {
+      *result = NULL;
+      return WindowsError(fname, GetLastError());
+    } else {
+      *result = new WindowsWritableFile(fname, h);
+      return Status::OK();
+    }
+  }
+
+  virtual bool FileExists(const std::string& fname) {
+    return _waccess(Utf8ToUtf16(fname).data(), 0) == 0;
+  }
+
+  virtual Status GetChildren(const std::string& dir,
+                             std::vector<std::string>* result) {
+    result->clear();
+
+    WIN32_FIND_DATAW data;
+    HANDLE handle = FindFirstFileW(Utf8ToUtf16(dir + "/*").data(), &data);
+    if (handle == INVALID_HANDLE_VALUE) {
+      return WindowsError(dir, GetLastError());
+    }
+    while (FindNextFileW(handle, &data) != 0) {
+      std::vector<char> v = Utf16ToUtf8(data.cFileName);
+      if (!(strcmp(v.data(), ".") == 0 || strcmp(v.data(), "..") == 0)) {
```

Comment:
Any particular reason to have this extra check here that drops "." and ".." and then have to deal with the difference from Posix in one of the tests you updated?